torch/testing/_internal/common_methods_invocations.py

```diff
@@ -4402,6 +4418,20 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
            supports_inplace_autograd=False,
            sample_inputs_func=sample_inputs_kron),
+    OpInfo('inner',
+           dtypes=floating_types(),
+           dtypesIfCPU=all_types_and_complex_and(torch.float16, torch.bfloat16),
+           # BFloat16 support on CUDA requires CUDA 11 and SM53
+           dtypesIfCUDA=floating_types_and(torch.float16, torch.complex64, torch.complex128,
+                                           *[torch.bfloat16] if CUDA11OrLater else []),
+           dtypesIfROCM=floating_types_and(torch.half),
```

Comment:
I don't see `inner` in `derivatives.yaml`. Does that mean it doesn't support autograd?

Reply:
Hey @imaginary-person! That's a good question, and we have to look at native_functions.yaml to know the answer:

https://github.com/pytorch/pytorch/blob/85126629a59ac51dcebca08b0ef2da5dd3afd827/aten/src/ATen/native/native_functions.yaml#L8730

Here we see that inner has no `DISPATCH` entry in its yaml, which means it acquires the default `DISPATCH` type of CompositeImplicitAutograd (see the README [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md)). 

Basically, we can think of "CompositeImplicitAutograd" as letting inner's autograd formula be defined by the operations its composed of, so it doesn't need a native_functions.yaml entry to support autograd.

"CompositeExplicitAutograd," by the way, means the operation is still composite but implements its own autograd. So all functions that support autograd either have derivatives.yaml entries or are CompositeImplicitAutograd dispatch types.

Does that make sense? 

Now as for testing, it's true that there may be some cases for complicit operations where testing there behavior is redundant with testing the behavior of the ops they're composed of. However, determining this would take some more work, and it would be tricky to keep up to date with operator implementations. If the implementation of inner changed, for example, maybe we would need to carefully tweak its test surface again. So for now we try to be (mostly) agnostic about how operators are implemented OR, if we do have implementation-specific tests, we try to make them fail if an operator's implementation changes.

Reply:
Thanks a lot for the clarification, @mruberry!